SpatialGDK/Source/SpatialGDKServices/Private/SpatialPackageManager.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Async/Async.h"
+#include "SpatialPackageManager.h"
+#include "SpatialGDKServicesConstants.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialPackageManager);
+
+FSpatialPackageManager::FSpatialPackageManager() {}
+
+void FSpatialPackageManager::FetchRuntimeBinary(FString RuntimeVersion)
+{
+	FString RuntimePath = FPaths::Combine(SpatialGDKServicesConstants::GDKProgramPath, TEXT("runtime"), RuntimeVersion);
+
+	// Check if the binary already exists for a given version
+	if (FPaths::FileExists(FPaths::Combine(RuntimePath, SpatialGDKServicesConstants::RuntimeExe)))
+	{
+		UE_LOG(LogSpatialPackageManager, Verbose, TEXT("Runtime binary already exists."));
+		return;
+	}
+
+	// If it does not exist then fetch the binary using `spatial worker package retrieve`
+	UE_LOG(LogSpatialPackageManager, Log, TEXT("Trying to fetch runtime version %s"), *RuntimeVersion);
+
+	// Download and unzip the executable to "UnrealGDK\SpatialGDK\Binaries\ThirdParty\Improbable\Programs\Runtime\*version*"
+	FString Params =
+		FString::Printf(TEXT("package retrieve runtime %s %s %s --unzip"), *SpatialGDKServicesConstants::PlatformVersion, *RuntimeVersion, *RuntimePath);
+	FSpatialPackageManager FSpatialPackageManager;
+	FSpatialPackageManager.StartProcess(Params, "Runtime Fetching");
+}
+
+void FSpatialPackageManager::FetchInspectorBinary(FString InspectorVersion)
+{
+	FString InspectorPath = FPaths::Combine(SpatialGDKServicesConstants::GDKProgramPath, TEXT("inspector"), InspectorVersion);
+
+	// Check if the binary already exists
+	if (FPaths::FileExists(FPaths::Combine (InspectorPath, SpatialGDKServicesConstants::InspectorExe)))
+	{
+		UE_LOG(LogSpatialPackageManager, Verbose, TEXT("Inspector binary already exists."));
+		return;
+	}
+
+	// If it does not exist then fetch the binary using `spatial worker package get`
+	// Download the package to UnrealGDK\SpatialGDK\Binaries\ThirdParty\Improbable\Programs
+	FString Params = FString::Printf(TEXT("package get inspector %s %s %s/%s"), *SpatialGDKServicesConstants::PlatformVersion, *InspectorVersion,
+									 *InspectorPath, *SpatialGDKServicesConstants::InspectorExe);
+	FSpatialPackageManager FSpatialPackageManager;
+	FSpatialPackageManager.StartProcess(Params, "Inspector Fetching");
+}
+
+void FSpatialPackageManager::StartProcess(FString Params, FString ProcessName)
+{
+	auto ExePath = SpatialGDKServicesConstants::SpatialExe;
+	FetchingProcess = {ExePath, Params, true, true };
+	FetchingProcess->OnOutput().BindLambda([&](const FString& Output) {
+		UE_LOG(LogSpatialPackageManager, Display, TEXT("Runtime: %s"), *Output);
+	});
+	FetchingProcess->Launch();
+
+	while (FetchingProcess->Update())
+	{
+		if (FetchingProcess->GetDuration().GetTotalSeconds() > SpatialGDKServicesConstants::ProcessTimeoutTime)
+		{
+			UE_LOG(LogSpatialPackageManager, Error, TEXT("Timed out waiting for the %s fetching to start."), *ProcessName);
```

Comment:
Minor thing: this would currently print "Timed out waiting for the Inspector/Runtime Fetching fetching to start.", right? I think we can just remove the " Fetching" part from the StartProcess calls.

Reply:
Ah, there could be a problem with runtime.exe and this Runtime process. The cleanest solution would be to kill the process by id rather than name, but leaving it up to you.